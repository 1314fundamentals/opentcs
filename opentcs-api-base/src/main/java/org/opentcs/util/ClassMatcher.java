// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.util;

import java.io.Serializable;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.function.Predicate;

/**
 * Checks whether the input argument is assignable to any of a given set of classes.
 */
public class ClassMatcher
    implements
      Predicate<Object>,
      Serializable {

  /**
   * The set of classes to evaluate incoming events to.
   */
  private final Set<Class<?>> clazzes = new HashSet<>();

  /**
   * Creates a new instance.
   *
   * @param events The set of classes to evaluate incoming events to.
   */
  public ClassMatcher(Class<?>... events) {
    clazzes.addAll(Arrays.asList(events));
  }

  @Override
  public boolean test(Object object) {
    return clazzes.stream().anyMatch(clazz -> clazz.isAssignableFrom(object.getClass()));
  }

  @Override
  public Predicate<Object> negate() {
    return new ClassMatcher() {
      @Override
      public boolean test(Object t) {
        return !ClassMatcher.this.test(t);
      }
    };
  }
}
