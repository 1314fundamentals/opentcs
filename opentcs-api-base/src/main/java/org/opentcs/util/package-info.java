// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
/**
 * A collection of classes providing miscellaneous functions that are used
 * throughout openTCS.
 */
package org.opentcs.util;
