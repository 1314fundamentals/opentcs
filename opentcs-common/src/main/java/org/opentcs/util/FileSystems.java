// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.util;

import static java.util.Objects.requireNonNull;

import java.io.File;

/**
 * This class provides helper methods for working with file systems.
 */
public final class FileSystems {

  /**
   * Prevents creation of instances.
   */
  private FileSystems() {
  }

  /**
   * Recursively deletes a given file/directory.
   *
   * @param target The file/directory to be deleted recursively.
   * @return <code>true</code> if deleting the target was successful, else
   * <code>false</code>.
   */
  public static boolean deleteRecursively(File target) {
    requireNonNull(target, "target");

    // If the target is a directory, remove its contents first.
    if (target.isDirectory()) {
      File[] entries = target.listFiles();
      for (File curEntry : entries) {
        boolean successful;

        if (curEntry.isDirectory()) {
          successful = deleteRecursively(curEntry);
        }
        else {
          successful = curEntry.delete();
        }

        if (!successful) {
          return false;
        }
      }
    }

    return target.delete();
  }
}
