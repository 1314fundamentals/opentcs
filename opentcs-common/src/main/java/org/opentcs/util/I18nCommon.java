// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.util;

/**
 * Defines constants regarding internationalization.
 */
public interface I18nCommon {

  /**
   * The path to the project's resource bundle.
   */
  String BUNDLE_PATH = "i18n/org/opentcs/common/Bundle";
}
