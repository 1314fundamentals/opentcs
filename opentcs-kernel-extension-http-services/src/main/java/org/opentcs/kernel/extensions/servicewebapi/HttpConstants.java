// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi;

/**
 * Defines some HTTP-related constants.
 */
public class HttpConstants {

  /**
   * Name of the header that is expected to contain the API access keys.
   */
  public static final String HEADER_NAME_ACCESS_KEY = "X-Api-Access-Key";
  /**
   * Content type for plain text.
   */
  public static final String CONTENT_TYPE_TEXT_PLAIN_UTF8 = "text/plain; charset=utf-8";
  /**
   * Content type for JSON structures.
   */
  public static final String CONTENT_TYPE_APPLICATION_JSON_UTF8 = "application/json; charset=utf-8";

  /**
   * Prevents instantiation.
   */
  private HttpConstants() {
  }
}
