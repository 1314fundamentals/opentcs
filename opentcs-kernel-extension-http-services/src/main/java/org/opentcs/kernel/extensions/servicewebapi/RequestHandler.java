// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi;

import io.javalin.apibuilder.EndpointGroup;
import org.opentcs.components.Lifecycle;

/**
 * A request handler.
 */
public interface RequestHandler
    extends
      Lifecycle {

  /**
   * Creates the handler's routes for the web service.
   */
  EndpointGroup createRoutes();
}
