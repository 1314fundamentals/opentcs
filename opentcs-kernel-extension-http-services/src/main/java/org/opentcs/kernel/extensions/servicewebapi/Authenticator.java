// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi;

import static java.util.Objects.requireNonNull;

import com.google.common.base.Strings;
import io.javalin.http.Context;
import jakarta.inject.Inject;
import java.util.Objects;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Authenticates incoming requests.
 */
public class Authenticator {

  /**
   * This class's logger.
   */
  private static final Logger LOG = LoggerFactory.getLogger(Authenticator.class);
  /**
   * Defines the required access rules.
   */
  private final ServiceWebApiConfiguration configuration;

  /**
   * Creates a new instance.
   *
   * @param configuration Defines the required access rules.
   */
  @Inject
  public Authenticator(ServiceWebApiConfiguration configuration) {
    this.configuration = requireNonNull(configuration, "configuration");
  }

  /**
   * Checks whether authentication is required and the given request is authenticated.
   *
   * @param context The context to be checked.
   * @return <code>true</code> if, and only if, authentication is required and the given request is
   * authenticated.
   */
  public boolean isAuthenticated(Context context) {
    requireNonNull(context, "context");

    String requestAccessKey = context.header(HttpConstants.HEADER_NAME_ACCESS_KEY);
    LOG.debug(
        "Provided access key in header is '{}', required value is '{}'",
        requestAccessKey,
        configuration.accessKey()
    );

    // Any empty access key indicates authentication is not required.
    if (Strings.isNullOrEmpty(configuration.accessKey())) {
      LOG.debug("No access key, authentication not required.");
      return true;
    }

    return Objects.equals(requestAccessKey, configuration.accessKey());
  }

}
