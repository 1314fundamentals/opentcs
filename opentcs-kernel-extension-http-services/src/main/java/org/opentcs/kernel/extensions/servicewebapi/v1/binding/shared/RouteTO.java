// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared;

import static java.util.Objects.requireNonNull;

import jakarta.annotation.Nonnull;
import jakarta.annotation.Nullable;
import java.util.List;

public class RouteTO {
  private long costs = -1;
  private List<Step> steps;

  public RouteTO(
      long costs,
      List<Step> steps
  ) {
    this.costs = costs;
    this.steps = requireNonNull(steps, "steps");
  }

  public long getCosts() {
    return costs;
  }

  public RouteTO setCosts(long costs) {
    this.costs = costs;
    return this;
  }

  public List<Step> getSteps() {
    return steps;
  }

  public RouteTO setSteps(List<Step> steps) {
    this.steps = steps;
    return this;
  }

  public static class Step {

    private String path;
    private String sourcePoint;
    private String destinationPoint;
    private VehicleOrientationTO vehicleOrientation;
    private int routeIndex;
    private long costs;
    private boolean executionAllowed;
    private ReroutingTypeTO reroutingType;

    public Step(
        String path,
        String sourcePoint,
        String destinationPoint,
        VehicleOrientationTO vehicleOrientation,
        int routeIndex,
        long costs,
        boolean executionAllowed,
        ReroutingTypeTO reroutingType
    ) {
      this.path = path;
      this.sourcePoint = sourcePoint;
      this.destinationPoint = requireNonNull(destinationPoint, "destinationPoint");
      this.vehicleOrientation = requireNonNull(vehicleOrientation, "vehicleOrientation");
      this.routeIndex = routeIndex;
      this.costs = costs;
      this.executionAllowed = executionAllowed;
      this.reroutingType = reroutingType;
    }

    @Nullable
    public String getPath() {
      return path;
    }

    public Step setPath(String path) {
      this.path = path;
      return this;
    }

    @Nullable
    public String getSourcePoint() {
      return sourcePoint;
    }

    public Step setSourcePoint(String sourcePoint) {
      this.sourcePoint = sourcePoint;
      return this;
    }

    @Nonnull
    public String getDestinationPoint() {
      return destinationPoint;
    }

    public Step setDestinationPoint(
        @Nonnull
        String destinationPoint
    ) {
      this.destinationPoint = requireNonNull(destinationPoint, "destinationPoint");
      return this;
    }

    @Nonnull
    public VehicleOrientationTO getVehicleOrientation() {
      return vehicleOrientation;
    }

    public Step setVehicleOrientation(
        @Nonnull
        VehicleOrientationTO vehicleOrientation
    ) {
      this.vehicleOrientation = requireNonNull(vehicleOrientation, "vehicleOrientation");
      return this;
    }

    public int getRouteIndex() {
      return routeIndex;
    }

    public Step setRouteIndex(int routeIndex) {
      this.routeIndex = routeIndex;
      return this;
    }

    public long getCosts() {
      return costs;
    }

    public Step setCosts(long costs) {
      this.costs = costs;
      return this;
    }

    public boolean isExecutionAllowed() {
      return executionAllowed;
    }

    public Step setExecutionAllowed(boolean executionAllowed) {
      this.executionAllowed = executionAllowed;
      return this;
    }

    public ReroutingTypeTO getReroutingType() {
      return reroutingType;
    }

    public Step setReroutingType(ReroutingTypeTO reroutingType) {
      this.reroutingType = reroutingType;
      return this;
    }

    // CHECKSTYLE:OFF
    public enum VehicleOrientationTO {
      FORWARD,
      BACKWARD,
      UNDEFINED
    }

    public enum ReroutingTypeTO {
      REGULAR,
      FORCED
    }
    // CHECKSTYLE:ON
  }
}
