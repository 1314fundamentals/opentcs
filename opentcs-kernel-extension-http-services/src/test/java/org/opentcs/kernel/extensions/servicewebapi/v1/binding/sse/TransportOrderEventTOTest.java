// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.opentcs.kernel.extensions.servicewebapi.JsonBinder;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse.TransportOrderEventTO.DriveOrderTO.RouteTO.ReroutingTypeTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse.TransportOrderEventTO.DriveOrderTO.RouteTO.VehicleOrientationTO;

/**
 * Tests for {@link TransportOrderEventTO}.
 */
class TransportOrderEventTOTest {

  private JsonBinder jsonBinder;

  @BeforeEach
  void setUp() {
    jsonBinder = new JsonBinder();
  }

  @Test
  void jsonSample() {
    TransportOrderEventTO to = new TransportOrderEventTO(
        createTransportOrder(TransportOrderEventTO.StateTO.RAW),
        createTransportOrder(TransportOrderEventTO.StateTO.BEING_PROCESSED)
    );

    Approvals.verify(jsonBinder.toJson(to));
  }

  private TransportOrderEventTO.TransportOrderTO createTransportOrder(
      TransportOrderEventTO.StateTO state
  ) {
    Instant time = Instant.parse("2025-07-22T12:03:00Z");
    return new TransportOrderEventTO.TransportOrderTO()
        .setName("order-1")
        .setProperties(new TreeMap<>(Map.of("key-1", "value-1", "key-2", "value-2")))
        .setHistory(
            new ObjectHistoryTO().setEntries(
                List.of(
                    new ObjectHistoryTO.ObjectHistoryEntryTO()
                        .setEventCode("event-code-1")
                        .setTimestamp(time)
                        .setSupplements(List.of("supplement-1", "supplement-2"))
                )
            )
        )
        .setType("type-1")
        .setDependencies(List.of("dependency-1", "dependency-2"))
        .setDriveOrders(
            List.of(
                new TransportOrderEventTO.DriveOrderTO()
                    .setName("drive-order-1")
                    .setDestination(
                        new TransportOrderEventTO.DriveOrderTO.DestinationTO()
                            .setDestination("destination-1")
                            .setOperation("operation-1")
                            .setProperties(
                                new TreeMap<>(Map.of("key-3", "value-3", "key-4", "value-4"))
                            )
                    )
                    .setTransportOrder("order-1")
                    .setRoute(
                        new TransportOrderEventTO.DriveOrderTO.RouteTO()
                            .setSteps(
                                List.of(
                                    new TransportOrderEventTO.DriveOrderTO.RouteTO.StepTO()
                                        .setPath("path-1")
                                        .setSourcePoint("point-1")
                                        .setDestinationPoint("point-2")
                                        .setVehicleOrientation(VehicleOrientationTO.FORWARD)
                                        .setRouteIndex(1)
                                        .setCosts(100)
                                        .setExecutionAllowed(true)
                                        .setReroutingType(ReroutingTypeTO.REGULAR)
                                )
                            )
                            .setCosts(200)

                    )
                    .setState(TransportOrderEventTO.DriveOrderTO.StateTO.TRAVELLING)
            )
        )
        .setPeripheralReservationToken("reservation-token-1")
        .setCurrentDriveOrderIndex(3)
        .setCurrentRouteStepIndex(5)
        .setState(state)
        .setCreationTime(time)
        .setDeadline(time.plus(1, ChronoUnit.HOURS))
        .setFinishedTime(time.plus(30, ChronoUnit.MINUTES))
        .setIntendedVehicle("vehicle-1")
        .setProcessingVehicle("vehicle-2")
        .setWrappingSequence("sequence-1")
        .setDispensable(true);
  }
}
