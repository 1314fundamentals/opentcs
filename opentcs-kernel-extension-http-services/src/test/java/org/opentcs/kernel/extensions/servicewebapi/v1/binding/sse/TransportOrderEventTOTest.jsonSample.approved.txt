{
  "currentObjectState" : {
    "name" : "order-1",
    "properties" : {
      "key-1" : "value-1",
      "key-2" : "value-2"
    },
    "history" : {
      "entries" : [ {
        "timestamp" : "2025-07-22T12:03:00Z",
        "eventCode" : "event-code-1",
        "supplements" : [ "supplement-1", "supplement-2" ]
      } ]
    },
    "type" : "type-1",
    "dependencies" : [ "dependency-1", "dependency-2" ],
    "driveOrders" : [ {
      "name" : "drive-order-1",
      "destination" : {
        "destination" : "destination-1",
        "operation" : "operation-1",
        "properties" : {
          "key-3" : "value-3",
          "key-4" : "value-4"
        }
      },
      "transportOrder" : "order-1",
      "route" : {
        "steps" : [ {
          "path" : "path-1",
          "sourcePoint" : "point-1",
          "destinationPoint" : "point-2",
          "vehicleOrientation" : "FORWARD",
          "routeIndex" : 1,
          "costs" : 100,
          "executionAllowed" : true,
          "reroutingType" : "REGULAR"
        } ],
        "costs" : 200
      },
      "state" : "TRAVELLING"
    } ],
    "peripheralReservationToken" : "reservation-token-1",
    "currentDriveOrderIndex" : 3,
    "currentRouteStepIndex" : 5,
    "state" : "RAW",
    "creationTime" : "2025-07-22T12:03:00Z",
    "deadline" : "2025-07-22T13:03:00Z",
    "finishedTime" : "2025-07-22T12:33:00Z",
    "intendedVehicle" : "vehicle-1",
    "processingVehicle" : "vehicle-2",
    "wrappingSequence" : "sequence-1",
    "dispensable" : true
  },
  "previousObjectState" : {
    "name" : "order-1",
    "properties" : {
      "key-1" : "value-1",
      "key-2" : "value-2"
    },
    "history" : {
      "entries" : [ {
        "timestamp" : "2025-07-22T12:03:00Z",
        "eventCode" : "event-code-1",
        "supplements" : [ "supplement-1", "supplement-2" ]
      } ]
    },
    "type" : "type-1",
    "dependencies" : [ "dependency-1", "dependency-2" ],
    "driveOrders" : [ {
      "name" : "drive-order-1",
      "destination" : {
        "destination" : "destination-1",
        "operation" : "operation-1",
        "properties" : {
          "key-3" : "value-3",
          "key-4" : "value-4"
        }
      },
      "transportOrder" : "order-1",
      "route" : {
        "steps" : [ {
          "path" : "path-1",
          "sourcePoint" : "point-1",
          "destinationPoint" : "point-2",
          "vehicleOrientation" : "FORWARD",
          "routeIndex" : 1,
          "costs" : 100,
          "executionAllowed" : true,
          "reroutingType" : "REGULAR"
        } ],
        "costs" : 200
      },
      "state" : "TRAVELLING"
    } ],
    "peripheralReservationToken" : "reservation-token-1",
    "currentDriveOrderIndex" : 3,
    "currentRouteStepIndex" : 5,
    "state" : "BEING_PROCESSED",
    "creationTime" : "2025-07-22T12:03:00Z",
    "deadline" : "2025-07-22T13:03:00Z",
    "finishedTime" : "2025-07-22T12:33:00Z",
    "intendedVehicle" : "vehicle-1",
    "processingVehicle" : "vehicle-2",
    "wrappingSequence" : "sequence-1",
    "dispensable" : true
  }
}